doc/reference/trd104-syscalls.md

```diff
@@ -303,61 +303,92 @@ call to invoke. The kernel invokes upcalls only in response to Yield
 
-The yield number specifies which call is invoked.
+The Yield number (in r0) specifies which call is invoked:
 
 | System call     | Yield number value |
 |-----------------|--------------------|
 | yield-no-wait   |                  0 |
 | yield-wait      |                  1 |
-
+| yield-wait-for  |                  2 |
 
 All other yield number values are reserved. If an invalid
 yield number is passed the kernel MUST return immediately.
 
-The no wait field is only used by `yield-no-wait`. It contains the
-memory address of an 8-bit byte that `yield-no-wait` writes to
-indicate whether an upcall was invoked. If invoking `yield-no-wait`
-resulted in an upcall executing, `yield-no-wait` writes 1 to the
-field address. If invoking `yield-no-wait` resulted in no upcall
-executing, `yield-no-wait` writes 0 to the field address. This field
-allows userspace loops that want to flush the upcall queue to
-execute `yield-no-wait` until the queue is empty.
+The meaning of `yield-param-N` (r1-r3) is specific to the yield type.
+
+
+### 4.1.1 Yield-NoWait
+
+Yield number 0, Yield-NoWait, executes a single upcall if any is
+pending.  If no upcalls are pending it returns immediately.
+
+`yield-param-1` contains the memory address of an 8-bit byte that
+Yield-NoWait writes to indicate whether an upcall was invoked. If
+invoking Yield-NoWait resulted in an upcall executing, Yield-NoWait
+writes 1 to the field address. If invoking Yield-NoWait resulted in no
+upcall executing, Yield-NoWait writes 0 to the field address. This field
+allows userspace loops that want to flush the upcall queue to execute
+Yield-NoWait until the queue is empty.
+
+`yield-param-2` and `yield-param-3` are unused and reserved.
+
+
+### 4.1.2 Yield-Wait
+
+Yield number 1, Yield-Wait, blocks until an upcall executes. It is
+commonly used to provide a blocking I/O interface to userspace or to
+indicate to the kernel that the process has no work to do so the system
+can be put into a sleep state. A userspace library starts a long-running
+operation that has an upcall, then calls Yield-Wait to wait for an
+upcall.  When the Yield-Wait returns, the process checks if the resuming
+upcall was the one it was expecting, and if not calls Yield-Wait again.
+
+`yield-param-1`, `yield-param-2`, and `yield-param-3` are unused and reserved.
+
+
+### 4.1.3 Yield-WaitFor
+
+The third call, Yield-WaitFor, blocks until a specific upcall executes.
+In contrast to the Yield-Wait call, only the specified upcall is
+eligible to execute. Any other upcall that arrives will be queued.
+
+`yield-param-1` is the Driver number and `yield-param-2` is the
+Subscribe number to wait for.
+
+`yield-param-3` is unused and reserved.
+
+_Note:_ In the case that the specified upcall is set to the Null Upcall,
+no upcall will execute, but Yield-WaitFor will still return. In this
+case, the contents of r0-r3 upon return are UNDEFINED and should not be
+relied on in any way.
```

Comment:
OK, this is great except why not have `r0`-`r3` be the values that would have been passed a non-null upcall? This would allow many cases to avoid callbacks entirely and just process the results directly, and doesn't cost anything (I don't think).

Reply:
That is `yield-for-blocking`.



>  Disadvantages:
> 
>  Semantic confusion: schedule_upcall() in a capsule may not, actually, schedule an upcall any more. Instead the "upcall" happens as the return to a yield.
>  If userspace does call subscribe, but then only uses yield-for-blocking, that upcall would never get called, which is odd. But with a mix of yield-for-blocking and yield, the upcall would get called sometimes. This inconsistency complicates the interface.
>

Reply:
Okay, so, the reason I didn't do this is because I couldn't decide the right behavior for `r3`.

TRD104 starts by describing upcalls abstractly as something that have (up to) 4 return arguments passed in r0-r3. We have one type of upcall the kernel actually emits right now, namely the upcall produced in response to subscribe [n.b., nothing in this TRD forbids other upcall signatures in the future*]. For clarity, I will use `SubscribableUpcall` to refer to an upcall invoked on a pointer passed via Subscribe.

Much to my surprise, as far as I can tell, the signature of `SubscribableUpcall` is not expressly defined anywhere except in [§5.2 of this TRD, where the C prototype for a `subscribe_upcall` is given as an example](https://github.com/tock/tock/blob/9d63213714f41ba1a8c57e6981688c867c08922f/doc/reference/trd104-syscalls.md#52-subscribe).

> Indeed, [we define the `FunctionCall` object passed to UKB](https://docs.tockos.org/kernel/process/struct.functioncall) with this note:
> 
> Struct that defines a upcall that can be passed to a process. The upcall takes four arguments that are Driver and upcall specific, so they are represented generically here.

In practice, the type signature for an upcall is established by the kernel in [kernel::upcall::schedule()](https://github.com/tock/tock/blob/9d63213714f41ba1a8c57e6981688c867c08922f/kernel/src/upcall.rs#L133-L139) and an implicit assumption in the [definition/creation of an `Upcall` object](https://github.com/tock/tock/blob/9d63213714f41ba1a8c57e6981688c867c08922f/kernel/src/upcall.rs#L78-L79) that upcalls come only from subscribes. But that's just an implementation artifact at the moment, not part of our ABI*.

Now, if `YieldFor` were to say something like 'sets r0-r3 to the Upcall Arguments if the current Upcall is the NullUpcall', then we would expect all Upcall Arguments to be set, right? Now, for the case of a `SubscribableUpcall` r0-r2 are clear, but what do we do with r3, which would be the `userdata` pointer? Does `YieldFor` need to do different things based on which type of Upcall it's passing?

We could further amend TRD104 to formally specify the function signature for _all_ upcalls, and then define `YieldFor` to skip r3. Or define `SubscribableUpcall` and specify `YieldFor` behavior only for that upcall type.

I was shooting for minimum delta with first RFC, and I don't think we can do this "skip the callback" without also making more changes around Upcall definition. I lean to formalizing `SubscribableUpcall` as a no-overhead option for today (since there is only one Upcall type still) while leaving the door open to other Upcall signatures in the future.

I do think the "skip the callback" would be nice. I can update this RFC to capture a version of that.

---

*Subscribe does start like this:

> The Subscribe system call class is how a userspace process registers upcalls with the kernel. 

Which if we really read into the "a" there maybe suggests the 1:1 mapping, but I wouldn't blink an eye at a future section that said something like "The `GetASignal` system call is another way userspace processes register upcalls with the kernel". If there really is _OneTrueUpcallSignature_, this TRD needs to say that much more directly.

Reply:
> *Subscribe does start like this:
> 
> > The Subscribe system call class is how a userspace process registers upcalls with the kernel.
> 
> Which if we really read into the "a" there maybe suggests the 1:1 mapping, but I wouldn't blink an eye at a future section that said something like "The `GetASignal` system call is another way userspace processes register upcalls with the kernel". If there really is _OneTrueUpcallSignature_, this TRD needs to say that much more directly.

But wouldn't we need to add that everywhere? A command syscall is THE ONLY way to issue a command. Yield is THE ONLY way to yield, etc. It seems clear to me that subscribe is the only way to subscribe, since that is all it does.

But yes we need to document the upcall signature.

Reply:
I agree it's clear that "subscribe is the only way to subscribe".

What I do not think is clear is that "subscribe is the only way to get upcalls". Upcalls are defined completely independently of subscribe in the previous section. As written, subscribe is just one thing (and currently happens to be the only thing) than can generate an upcall.

Reply:
I don't think we need to say anything about subscribe being the only way to set upcalls, any more than we need to say command is the only way to instruct the kernel to start an operation. Commands are documented elsewhere. But, we do need to define what an upcall is, in my opinion.